### DISCUSSION TITLE
Providing support and tracking feature status for LLVM/Clang. libc++ (CPU), Clang vs Nvidia's CUDA compiler compatibility (OpenCL/GPGPU)

### DISCUSSION THREAD

**Discussion by advancedwebdeveloper** [onnx/onnx#3001] [Ideas]
@prasanthpul pointed me an an option of creating an issue - so I am opening one.
I am currently interested to get some testing and patch for ONNX on Windows 10 (32bit, 64bit) and Ubuntu 20 Linux (32bit, 64bit).
My extended interest assumes integration with https://nifi.apache.org/ and https://github.com/apache/nifi-minifi-cpp .

@tspannhw might suggest some cross-compilation scenarios and practical end-user use cases, for embedded boards.

I will open an issue on some bugs, which I am experiencing with protobuf - ONNX does not compile on Windows 10. But it does compile on Linux.
I will check specific versions of protobuf and reference protobuf's github issue with future comments to this discussion.

Computing on CPUs assume [Clang compiler](https://clang.llvm.org/) being capable of building against [libc++ standard library](https://libcxx.llvm.org/) . I am not sure about project specifics of ONNX and migration to more bleeding edge C++ standard (C++17...C++20) - so it would be appreciated if committers/maintainers would elaborate current/future plans.
Current status for libc++:
[libc++ C++14 Status](https://libcxx.llvm.org/cxx1y_status.html)
[libc++ C++17 Status](https://libcxx.llvm.org/cxx1z_status.html)
[libc++ C++20 Status](https://libcxx.llvm.org/cxx2a_status.html)
. Current status for Clang:
[C++ Support in Clang](https://clang.llvm.org/cxx_status.html)
.
I am targeting [LLVM 11.0.0 RC2](https://github.com/llvm/llvm-project/releases/tag/llvmorg-11.0.0-rc2)  
[It appears the newest version ](http://llvm.1065342.n5.nabble.com/llvm-dev-11-0-0-Release-Release-Candidate-2-is-here-td140542.html) .

As for CUDA - there is [some](https://llvm.org/docs/CompileCudaWithLLVM.html) support for Clang's front-end.
It seems that CMake has [some](https://gitlab.kitware.com/cmake/cmake/-/issues/16586) support here - so perhaps @tambry could explain in detail.

Ivan

**Comment by tambry** [comment]
>It seems that CMake has some support here - so perhaps @tambry could explain in detail.

CMake for CUDA+Clang support is as follows:
* Supported on Linux.
* [Windows support](https://gitlab.kitware.com/cmake/cmake/-/issues/20751) I had mostly working, but last few problems require significant refactoring. Not before CMake 3.20.
* Both official and scattered (distro/system) CUDA installations supported.
* Both libc++ and libstdc++ should work. Newer libstdc++ requires patch by distribution, needed to be [updated for version 10](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=968672).
* Separable compilation not supported. [CMake issue](https://gitlab.kitware.com/cmake/cmake/-/issues/20726). Got some work done on this, might pick it up and get it done in time for CMake 3.19.
* No 32-bit support, but that's also not supported for NVCC. Might add support for NVCC as it seemed simple enough.

My real-world scenario is CMake 3.18 and Clang 12 trunk with libc++ crosscompiling for NVIDIA Tegra Xavier using CUDA 10.2.19.  
Let me know if you encounter issues or have more specific questions. I'm open for consulting.

**Comment by advancedwebdeveloper** [comment]
@tambry , would get back to this later - I appreciate you extended explanation.

As I promised - here is a bug with ONNX's benchmarks (CPU related, I suppose):
[Missing proper includes for WrappedMutex and Mutex](https://github.com/protocolbuffers/protobuf/issues/7873)
[Unknown type names "WrappedMutex" and "Mutex", while building ONNX, using Clang 11 (Windows 10)](https://groups.google.com/g/protobuf/c/GLFWU3ZkvOM)
. It looks like it could be possible to use [libc++](https://libcxx.llvm.org/) on Windows 10.

Ivan

**Comment by prasanthpul** [comment]
@askhade can you comment on this?

**Comment by askhade** [comment]
@advancedwebdeveloper : 
Let me try an respond to these 1 by 1
Providing support and tracking feature status for LLVM/Clang - Based on your description it seems you are seeing issues when compiling onnx on windows with Clang right? whereas it succeeds on Linux. Have you tried MSVC? I am not sure whether Clang on windows for onnx is well tested so yes there could be bugs there. 

libc++ (CPU) - What is the exact ask here? Is it just understanding the current plan to migrate to the latest standard? Right now there is no plan but I think this is a good point... Right now ONNX uses C++11 and it is worth it to explore whether we can move to a later standard.  

Clang vs Nvidia's CUDA compiler compatibility (OpenCL/GPGPU): How is this related to ONNX builds?

Regarding the bugs you created - I will go over them and try to prioritize them, however as you know ONNX thrives on community contributions and you are welcome to send a PR to fix these bugs.

**Reply by advancedwebdeveloper** [reply]
> Providing support and tracking feature status for LLVM/Clang - Based on your description it seems you are seeing issues when compiling onnx on windows with Clang right? whereas it succeeds on Linux. Have you tried MSVC? I am not sure whether Clang on windows for onnx is well tested so yes there could be bugs there.

That is exactly what I was targeting: I am heading to track down various bugs, to see where Clang itself (the compiler) could be improved/patched.

> 
> libc++ (CPU) - What is the exact ask here? Is it just understanding the current plan to migrate to the latest standard? Right now there is no plan but I think this is a good point... Right now ONNX uses C++11 and it is worth it to explore whether we can move to a later standard.

Most of the open source software, implemented in C++, seems to be adopting for [libc++'s C++14 status](http://libcxx.llvm.org/cxx1y_status.html) . But farther elevation of supported C++ standard is possible - but not fully supported by libc++, as for now:
[C++17 status, for libc++](http://libcxx.llvm.org/cxx1z_status.html)
[http://libcxx.llvm.org/ts1z_status.html](http://libcxx.llvm.org/ts1z_status.html)
[https://libcxx.llvm.org/docs/FeatureTestMacroTable.html#feature-status](https://libcxx.llvm.org/docs/FeatureTestMacroTable.html#feature-status)
. Your team can create a PR and share the status of C++ features, which are expected to supported.
Would be interested to contribute by testing.

> 
> Clang vs Nvidia's CUDA compiler compatibility (OpenCL/GPGPU): How is this related to ONNX builds?
I am just curious how Clang's support of CUDA related stuff could be improved.
Tests are required.

> 
> Regarding the bugs you created - I will go over them and try to prioritize them, however as you know ONNX thrives on community contributions and you are welcome to send a PR to fix these bugs.

OK
